scriptworker/worker.py

```diff
@@ -90,6 +94,87 @@
     return status
 
 
+class RunTasks:
+    """Manages processing of Taskcluster tasks."""
+
+    def __init__(self):
+        """Constructor."""
+        self.future = None
+        self.task_process = None
+        self.is_cancelled = False
+
+    async def invoke(self, context):
+        """Claims and processes Taskcluster work.
+
+        Args:
+            context (scriptworker.context.Context): context of worker
+
+        Returns: status code of build
+
+        """
+        try:
+            # Note: claim_work(...) might not be safely interruptible! See
+            # https://bugzilla.mozilla.org/show_bug.cgi?id=1524069
+            tasks = await self._run_cancellable(claim_work(context))
+            if not tasks or not tasks.get('tasks', []):
+                await self._run_cancellable(asyncio.sleep(context.config['poll_interval']))
+                return None
+
+            # Assume only a single task, but should more than one fall through,
+            # run them sequentially.  A side effect is our return status will
+            # be the status of the final task run.
+            status = None
+            for task_defn in tasks.get('tasks', []):
+                prepare_to_run_task(context, task_defn)
+                reclaim_fut = context.event_loop.create_task(reclaim_task(context, context.task))
+                status = await do_run_task(context, self._cancellable_verify_chain_of_trust, self._to_cancellable_process)
+                status = worst_level(status, await do_upload(context))
```

Comment:
Does this mean that if we cancel the task during `do_run_task` we don't upload the logs at all?
We may want to change this so we set a flag on upload. If we hit `CancelledError` and this flag isn't set, we upload the minimum: possibly `cot.log` and `live_backing.log` if they exist; probably skip cot generation + signing?

Reply:
I'm not sure I fully understand.
FWIW, our cancellable states are:
* `claim_work`: When cancelled, we jump to `except CancelledError` and return
* `sleep`: When cancelled, we jump to `except CancelledError` and return
* `verify_chain_of_trust`: When cancelled, skip running the task itself. Upload artifacts to Taskcluster and inform of `worker-shutdown` status (note that the `CancelledError` is caught in `do_run_task`!)
* Task process (within `run_task`): When cancelled, kill the process. Upload artifacts to Taskcluster and inform of `worker-shutdown` status (note that `CancelledError` is not raised! The process is killed, which triggers the follow-up behaviour)

I agree that perhaps the post-task behaviour should perhaps be different if the task was cancelled partially through a build - perhaps just: {upload `cot.log`, upload `live_backing.log`, set task status}, yeah. Should that be handled here, or in follow-up? I'm flexible :)

Reply:
I think you understand with the last paragraph. I basically meant that, unless I'm misreading, a cancel in the middle of the task will not hit the `do_upload` call. We may want an upload call after a cancel.

Maybe the upload call post-cancel should be a best effort upload:
* allow for missing artifacts (one of the logs may not exist?)
* allow for already-uploaded artifacts, in case we were killed mid- or post- upload

I'm flexible as well, since I think current behavior is `claim-expired` up to 20 min later with no logs. If the rollout process is lengthy enough for this change, it might be worth having both changes landed at the same time, but that's not a hard requirement.

Reply:
Ah, gotcha!
With the implementation in this PR:
* If there's a cancel in the middle of a task, `do_upload` **is** still called
* `do_upload` can't be cancelled. Only those four states (`claim_work`, `sleep`, `verify_chain_of_trust` and the child task process can be cancelled)

Reply:
Ok, misread, I thought we raised a `CancelledTask` which would jump to the `except`.
If we run `do_upload` regardless, we may want to change its behavior based on whether we cancelled (don't upload artifacts other than logs)

Reply:
Ah, yeah, very much on the same page!

I'm less confident that the cost-vs-benefit is worth it for attempting to add artifacts up until being killed - if the job failed due to `worker-shutdown`, and is being rescheduled to be done properly, I think that the potentially-incomplete/corrupt artifacts won't have much value. Besides, once we start dancing purposely close to being killed, it becomes tougher (what if we're killed in the middle of uploading? Are our operations atomic?).

Eh, as you said, this would be follow-up/discussion for later anyways :)

Reply:
> I'm less confident that the cost-vs-benefit is worth it for attempting to add artifacts up until being killed

This already happens: signingscript copies artifacts into `artifact_dir` as they're ready, and `do_upload` uploads everything in `artifact_dir`. We'd have to add logic to not upload the artifacts if that risks hitting the 30s mark. Potential followup, yes.

Reply:
I've adjusted the behaviour such that we only upload `chain_of_trust.log` and `live_backing.log` if the worker shuts down :+1: